backends/backends.c

```diff
@@ -246,6 +246,114 @@ static void backends_main_cleanup(void *ptr) {
     static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
 }
 
+/**
+ * Set type
+ *
+ * Set the variables necessaries to work with a specific backend.
+ *
+ * @param default_port  the default port of the backend
+ * @param backend_response_checker function called to check the result.
+ * @param backend_request_formatter function called to format the msessage to the backend
+ * @param type the backend string selector.
+ *
+ * @return It returns 0 case there is not necessity to set additional variables, 1 case it is necessary to adjust prometheus
+ *          2 case it is necessary to adjust kinesis and -1 on error
+ */
+int backend_set_type(int *default_port
+                     ,int (**backend_response_checker)(BUFFER *)
+                     ,int (**backend_request_formatter)(BUFFER *, const char *, RRDHOST *, const char *, RRDSET *, RRDDIM *, time_t, time_t, BACKEND_OPTIONS)
+                     ,const char *type) {
```

Comment:
Hey @thiagoftsm 	

This is nice, but i think it would be better to move `type` parsing logic outside the func, it will simplify things

something like this

```c
enum backend_type {
    UNKNOWN,
    TYPE1,
    TYPE2,
};

enum backend_type parse_backend_type(const char *type);

int is_backend_type_https(enum backend_type);
/*
use it^^ instead magic thing `*(type+13) == 's'`
*/
```

After it you have simple `case ...switch` inside your `backend_set` function. 

Will this work? What do you think?

Reply:
Hi @ilyam8 , I agree this will work and it also looks like a good idea, but let me explain firstly the motive I wrote this way. As you can see I moved inside the function the common algorithm, but I was forced to use an integer return, because we have prometheus and kinesis doing more than the common. Case we move type out of the function, we would test the results to have the value compatible with would not avoid the treatment of not common variables in another place, no less important in the way that I wrote I do not need to test again with an "if/else if" or switch the result to define the common variables. If I understood correct, the current algorithm tests only two times to define the not common for the "normal" case, while the enum suggestion could force us to test 1 more time to let the code more readble.

Reply:
I can only repeat - do not optimize things until there is a reason for it. If you are in doubt do benchmark and see if it worth. Otherwise prefer readable over unreadable.

i am sure - case on enum is super mega fast and it costs nothing.

> but I was forced to use an integer return, because we have prometheus and kinesis doing more than the common

That means that something is wrong there and the function a candidate for refactoring.

___
That is how i think:
 - i see function name 'do_this'
 - i read function and it `does this, does that and that`

> backend_set_type

The function is not responsible for parsing, it only sets. Parse loginc should be outside. Or name it `backend_parse_type_and_set`. Looks bad, right? :smile: